src/NodeCompiler/CompilerContext.php

```diff
@@ -13,26 +14,38 @@ class CompilerContext
     private $reflector;
 
     /**
-     * @var ReflectionClass|null
+     * @var ReflectionClass|ReflectionFunction|null
      */
-    private $self;
+    private $context;
 
-    public function __construct(Reflector $reflector, ?ReflectionClass $self)
+    /**
+     * @param Reflector $reflector
+     * @param ReflectionClass|ReflectionFunction|null $context
+     * @throws \RuntimeException
+     */
+    public function __construct(Reflector $reflector, $context)
     {
         $this->reflector = $reflector;
-        $this->self = $self;
+
+        if ($context instanceof ReflectionClass || $context instanceof ReflectionFunction || $context == null) {
```

Comment:
Hmm, it looks like this PR actually introduces two new features:

 * `__CLASS__` and `__NAMESPACE__` support
 * Allowing context to be a `ReflectionFunction` (which is required to make `__NAMESPACE__` work for a function, as I understand)

Maybe separate them? (especially as the latter could potentially break BC)